File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0538_Convert_BST_to_Greater_Tree.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0538_Convert_BST_to_Greater_Tree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int sum = 0;

    public TreeNode convertBST(TreeNode root) {
        if (root == null) {
            return root;
        }
        sum = 0;
        help(root);
        return root;
    }

    private void help(TreeNode root) {
        if (root == null) {
            return;
        }
        help(root.right);
        sum += root.val;
        root.val = sum;
        help(root.left);
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0538_Convert_BST_to_Greater_Tree.cc

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode* convertBST(TreeNode* root) {
    if (!root) {
      return root;
    }
    int sum = 0;
    help(root, sum);
    return root;
  }

  void help(TreeNode* root, int& sum) {
    if (!root) {
      return;
    }
    help(root->right, sum);
    sum += root->val;
    root->val = sum;
    help(root->left, sum);
  }
};
